### PR TITLE
suggest fixing queries with ^ or $ in impossible places

### LINKED ISSUE #4625: suggest escapes for impossible queries involving ^ and $
#4604 showed how a query like `err$0` can make indexed search time out. It would be better to notify the user that the query cannot match anything and suggest a plausible variant of it that would match something.

### PR DESCRIPTION
**Pull request by ijt**
Fixes #4625 

After:
Screen Shot 2019-06-22 at 10 38 44

Before:
Screen Shot 2019-06-22 at 10 41 04

Test plan: unit test

### CODE REVIEW

File: cmd/frontend/graphqlbackend/search_results.go [hunk 1]
```diff
@@ -481,8 +481,29 @@ func (r *searchResolver) Results(ctx context.Context) (*searchResultsResolver, e
 // resultsWithTimeoutSuggestion calls doResults, and in case of deadline
 // exceeded returns a search alert with a did-you-mean link for the same
 // query with a longer timeout.
-func (r *searchResolver) resultsWithTimeoutSuggestion(ctx context.Context) (*searchResultsResolver, error) {
+func (r *searchResolver) resultsOrSuggestions(ctx context.Context) (*searchResultsResolver, error) {
 	start := time.Now()
+
+	qs := r.query.Syntax
+	qStr := qs.String()
+	qs.EscapeImpossibleCaretsDollars()
```
Comment (keegancsmith) [inline cmd/frontend/graphqlbackend/search_results.go]: this is modify the query in r.query. Is that safe from a concurrency perspective? Is that intentional to change the query for other methods?

File: cmd/frontend/internal/pkg/search/query/syntax/query.go [hunk 2]
```diff
@@ -27,6 +27,56 @@ func (q *Query) WithErrorsQuoted() *Query {
...
+				s = initialCaretRx.ReplaceAllString(s, `\^`)
+			}
+			if i+1 < len(q.Expr) {
+				s = finalUnescapedDollarRx.ReplaceAllString(s, `$1\$$`)
+			}
+			return s
+		}
+		// escape is called twice to handle for example `^^^` which would
+		// otherwise end up as `^\^^` whereas we want it to be escaped as
+		// `^\^\^`.
```
Comment (keegancsmith) [inline cmd/frontend/internal/pkg/search/query/syntax/query.go]: can you make your example `^^^^^`? This comment makes it seem like it wouldn't handle more than three.

File: cmd/frontend/internal/pkg/search/query/syntax/query_test.go [hunk 3]
```diff
@@ -55,6 +55,58 @@ func TestExpr_String(t *testing.T) {
...
+		{in: "^$", want: "^$"},
+		{in: "^^$$", want: `^\^\$$`},
+		{in: "^^^$$$", want: `^\^\^\$\$$`},
+		{in: "$^", want: `\$\^`},
+		{in: "$ ^", want: `\$ \^`},
+		{in: "a$", want: "a$"},
+		{in: "a$b", want: `a\$b`},
+		{in: "^a$", want: "^a$"},
+		{in: "a^b", want: `a\^b`},
+		{in: "a^$b a^$b", want: `a\^\$b a\^\$b`},
```
Comment (keegancsmith) [inline cmd/frontend/internal/pkg/search/query/syntax/query_test.go]: can you add test cases for `(^a|b)` and `[^$]`. Both are valid and useful regexes, but I believe your code may incorrectly report this as impossible.

Comment (ijt) [reply cmd/frontend/internal/pkg/search/query/syntax/query_test.go]: I think the cases you pointed out mean the complexity of implementing this correctly isn't going to be worth it. Closing this out.

**Pull request closed without merging**